extensions/pkg/terraformer/state.go

```diff
@@ -164,33 +164,62 @@ func sniffJSONStateVersion(stateConfigMap []byte) (uint64, error) {
 }
 
 // Initialize implements StateConfigMapInitializer
-func (f StateConfigMapInitializerFunc) Initialize(ctx context.Context, c client.Client, namespace, name string) error {
-	return f(ctx, c, namespace, name)
+func (f StateConfigMapInitializerFunc) Initialize(ctx context.Context, c client.Client, namespace, name string, ownerRef *metav1.OwnerReference) error {
+	return f(ctx, c, namespace, name, ownerRef)
 }
 
 // CreateState create terraform state config map and use empty state.
-// It does not create or update state ConfigMap if already exists.
-func CreateState(ctx context.Context, c client.Client, namespace, name string) error {
+// It does not create or update state ConfigMap if already exists,
+// but it tries to patch with owner reference.
+func CreateState(ctx context.Context, c client.Client, namespace, name string, ownerRef *metav1.OwnerReference) error {
 	configMap := &corev1.ConfigMap{
 		ObjectMeta: metav1.ObjectMeta{Namespace: namespace, Name: name},
 		Data: map[string]string{
 			StateKey: "",
 		},
 	}
 
-	if err := c.Create(ctx, configMap); err != nil && !apierrors.IsAlreadyExists(err) {
+	if ownerRef != nil {
+		configMap.OwnerReferences = []metav1.OwnerReference{
+			*ownerRef,
+		}
+	}
+
+	err := c.Create(ctx, configMap)
+	if err == nil {
+		return nil
+	}
+	if !apierrors.IsAlreadyExists(err) {
+		return err
+	}
+
+	if ownerRef == nil {
+		return nil
+	}
+
+	// patch the configMap with ownerReference, TODO: remove after several releases
+	if err := c.Get(ctx, client.ObjectKey{Namespace: namespace, Name: name}, configMap); err != nil {
 		return err
 	}
 
-	return nil
+	oldConfigMap := configMap.DeepCopy()
+	configMap.OwnerReferences = []metav1.OwnerReference{
+		*ownerRef,
+	}
+	return c.Patch(ctx, configMap, client.MergeFrom(oldConfigMap))
```

Comment:
This code will never be executed for already existing shoots with non-empty state, because `config.InitializeState` is `false` here https://github.com/gardener/gardener/blob/ea37e4c21fae2d255928ce6dbf301e57e1292e68/extensions/pkg/terraformer/config.go#L171-L175 , thus the `Initialize` call of the `CreateState` StateConfigMapInitializer will be skipped.

The alternative options that I see are
- get it like it is now and without ownerRef for existing state configmaps
- move the owner reference ensurer to the https://github.com/gardener/terraformer, although it is not the controller responsible for the creation of the resources. 
- the other option is to add completely independent ownerRef ensurer somewhere here https://github.com/gardener/gardener/blob/ea37e4c21fae2d255928ce6dbf301e57e1292e68/extensions/pkg/controller/infrastructure/reconciler.go#L135-L145 

@gardener/gardener-maintainers any feedback is welcomed on how to proceed with this change.

Reply:
Hm, what about doing it in the `execute` function  and remove it after several releases again? https://github.com/gardener/gardener/blob/master/extensions/pkg/terraformer/terraformer.go#L189